### PR TITLE
Modifier keys often stuck down on Mac client

### PR DESCRIPTION
**Pull request by jpmcmu**
On MacOSX modifier keys will often stick. I was able to verify that modifier state in the Synergy client is correct and the system modifier state is incorrect. Furthermore, I was able to verify that an event is posted to the operating system for the key up event. 

After determining that Synergy is in the correct state, I attempted to fix this issue by asking the operating system what its modifier state was before every event was sent and then sending extra events to correct the systems modifier state if it was incorrect. This solution worked but after looking at the log of corrections, it became evident that this code was only required to correct mouse events. This was due to the fact that all key events pass the modifier state along with them. So the final solution was to remove the consistency check and add the modifier state to all mouse events.

This seems like an odd fix, but I have been unable to cause keys stick after these changes. 

Steps to cause a modifier key stick before these changes:
1. Hold down a modifier key (Shift key, etc)
2. Move mouse as rapidly as possible without stopping
3. While continuing to move the mouse rapidly, release the modifier key
4. Continue moving the mouse for another few seconds

Following the above steps, without the changes, I was able to cause a key stick around once in every three tries. With these changes I have not been able to get a key to stick.

### CODE REVIEW

File: src/lib/platform/OSXKeyState.cpp [hunk 1]
```diff
@@ -189,7 +189,7 @@ OSXKeyState::OSXKeyState(IEventQueue* events) :
 OSXKeyState::OSXKeyState(IEventQueue* events, synergy::KeyMap& keyMap) :
 	KeyState(events, keyMap)
 {
-	init();
+	init();a
```
Comment (interfect) [inline src/lib/platform/OSXKeyState.cpp]: This doesn't look particularly compile-able.

**Comment by interfect** [comment]
Thanks for working this out. Apparently if it works you will have fixed six years of annoyingness.

**Comment by nbolton** [comment]
@speaker Please create PR branch. You may need to fix OSXKeyState.cpp:192 (as @interfect points out, doesn't look compile-able).

**Comment by jpmcmu** [comment]
Removed the stray character. Should be good to go now.

**Comment by beporter** [comment]
List of issues that may be on the verge of being solved:

* Left mouse button gets stuck in right click mode on OS X client #3823
* Ctrl or shift keys randomly simulated while in IE #3194
* Command key stuck (Mac client, Windows server) #4057
* Arrow keys and ctrl+arrow causes screen rotation on Windows #4449
* Modifier state becomes out of sync on screen lock #4519
* _(And the grand-pappy of them all: Sticky modifier/meta keys (shift, ctrl, alt) #9)_

If this really works, somebody needs to give @jpmcmu an **actual** medal. I may be in shock right now. :scream:

**Comment by XinyuHou** [comment]
Could someone test this pull request from our nightly build?

http://synergy-project.org/nightly?filter=a597596

**Comment by nbolton** [comment]
@beporter Thanks for signal boosting. Looks like you got a hit with @hidden72 and @timpokorny but a miss with @jtbraun. We're considering if we should merge this fix in... looks fairly strong, what do you think? @XinyuHou

**Comment by jpmcmu** [comment]
This fix would only work for MacOSX clients (The fix itself is in a Mac specific portion of the code base). Based on the issue report, https://github.com/synergy/synergy/issues/4519, I think @jtbraun was testing this fix on a Win7 client / Linux server setup.

**Comment by jtbraun** [comment]
Confirmed that I was testing on a Win7 client, Ubuntu 14.04 server.

**Comment by beporter** [comment]
I can also confirm that I've had stuck keys occur on my Windows **host** machine (with a Mac Synergy client connected). The fix should ideally be cross-platform.

**Comment by jpmcmu** [comment]
The key sticking issues are caused by a number of different bugs. This pull request only fixes one of those bugs, which is either caused by Apple not correctly documenting how MacOSX tracks modifier state or a bug in their code. ( I am leaning towards a bug, it looks a lot like a race condition to me)

As an example I found another bug that causes the Ctrl key to stick off when Synergy switches between different clients. I was able to verify that is a bug in the server code, but that is as far as I have gone towards correcting that particular bug. 

What I am getting at is the "sticky key" issue is caused by a number of different bugs that will require multiple solutions/pull requests to fix.

**Comment by beporter** [comment]
I take back what I said about giving @jpmcmu a medal-- just one isn't going to be enough.

**Comment by nbolton** [comment]
@beporter Harsh ;) Sounds to me like @jpmcmu is a hero!

**Comment by nbolton** [comment]
OK, based on @jtbraun's comment (his issue is about Windows client), I think we should merge this PR into master.

**Comment by nbolton** [comment]
Either @XinyuHou or @speaker will do the merge.

**Comment by franciozzy** [comment]
I'm SO HAPPY to see this issue fixed. Testing 1.7.2 now.

**Comment by apiliguzov** [comment]
This bug was a major inconvenience for me. I am so glad to hear it is fixed. Going to upgrade and test today!

**Comment by chrishas35** [comment]
This is incredible!  Thank you for fixing this long standing annoyance!!

**Comment by scosant** [comment]
This... is... awesome! Thank you for fixing this bug!

**Comment by oriolrius** [comment]
I updated to Synergy 1.7.2 in server and client and I have the same problem. My server is a Fedora 20 and my Client is a Windows 7. Both of them using 2 screen in full-HD. So far today everything was working Ok I updated to last version and everything started failing. What can I do?

Thanks for your help.

**Comment by jsnyder** [comment]
Just chiming in to say thank you on this and the triple click fix.  Both of these have been driving me slightly insane for months.

**Pull request merged**